controller/tap/server.go

```diff
@@ -568,63 +580,72 @@ func (s *GRPCTapServer) hydrateEventLabels(ev *public.TapEvent) {
-	switch {
-	case err != nil:
+	res, err := s.resourceForIP(ip)
+	if err != nil {
 		return err
-	case pod == nil:
-		log.Debugf("no pod for IP %s", addr.PublicIPToString(ip))
-		return nil
-	default:
-		ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(pod, false)
-		podLabels := pkgK8s.GetPodLabels(ownerKind, ownerName, pod)
+	}
+
+	switch v := res.(type) {
+	case *corev1.Pod:
+		if v == nil {
+			log.Debugf("no pod found for IP %s", addr.PublicIPToString(ip))
+			return nil
+		}
+		ownerKind, ownerName := s.k8sAPI.GetOwnerKindAndName(v, false)
+		podLabels := pkgK8s.GetPodLabels(ownerKind, ownerName, v)
 		for key, value := range podLabels {
 			labels[key] = value
 		}
-		labels[pkgK8s.Namespace] = pod.Namespace
-		return nil
+		labels[pkgK8s.Namespace] = v.Namespace
+	case *corev1.Node:
+		labels[pkgK8s.Node] = v.Name
 	}
+	return nil
 }
 
-// podForIP returns the pod corresponding to a given IP address, if one exists.
+// resourceForIP returns the node or pod corresponding to a given IP address.
 //
-// If multiple pods exist with the same IP address, this may be because some
-// are terminating and the IP has been assigned to a new pod. In this case, we
-// select the running pod, if one currently exists. If there is a single pod
-// which is not running, we return that pod. Otherwise we return `nil`, as we
-// cannot easily determine which pod sent a given request.
-//
-// If no pods were found for the provided IP address, it returns nil. Errors are
-// returned only in the event of an error indexing the pods list.
-func (s *GRPCTapServer) podForIP(ip *public.IPAddress) (*corev1.Pod, error) {
+// First it checks if the IP corresponds to a Node's internal IP and returns the
+// node if that's the case. Otherwise it checks the running pods that match the
+// IP. If exactly one is found, it's returned. Otherwise it returns nil. Errors
+// are returned only in the event of an error searching the indices.
+func (s *GRPCTapServer) resourceForIP(ip *public.IPAddress) (runtime.Object, error) {
 	ipStr := addr.PublicIPToString(ip)
-	objs, err := s.k8sAPI.Pod().Informer().GetIndexer().ByIndex(podIPIndex, ipStr)
 
+	nodes, err := s.k8sAPI.Node().Informer().GetIndexer().ByIndex(ipIndex, ipStr)
 	if err != nil {
 		return nil, err
 	}
+	if len(nodes) == 1 {
+		log.Debugf("found one node at IP %s", ipStr)
+		return nodes[0].(*corev1.Node), nil
+	}
 
-	if len(objs) == 1 {
+	pods, err := s.k8sAPI.Pod().Informer().GetIndexer().ByIndex(ipIndex, ipStr)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(pods) == 1 {
 		log.Debugf("found one pod at IP %s", ipStr)
-		// It's safe to cast elements of `objs` to a `Pod`s here (and in the
-		// loop below). If the object wasn't a pod, it should never have been
-		// indexed by the indexing func in the first place.
-		return objs[0].(*corev1.Pod), nil
+		return pods[0].(*corev1.Pod), nil
 	}
 
-	for _, obj := range objs {
+	var singleRunningPod *corev1.Pod
+	for _, obj := range pods {
```

Comment:
Will this cause the dashboard to output non-deterministic graphs? E.g., on first load, if there is only one pod that is running, this will return a non-nil `singleRunningPod`. But on refreshing the dashboard if more than one pods are running, then the previous pod won't show up in the dashboard, right?

Now that we explicitly handle node internal IP, will it still be possible for multiple running pods to have the same IP? Is it bad to just issue a warning with the names of the pods if `len(pods) > 0`, and just `return nil, nil`?

Reply:
> Will this cause the dashboard to output non-deterministic graphs? E.g., on first load, if there is only one pod that is running, this will return a non-nil singleRunningPod. But on refreshing the dashboard if more than one pods are running, then the previous pod won't show up in the dashboard, right?

Yes. On the first load the pod would have been interpreted as a regular unmeshed pod, and on the second run that assumption would have been proven wrong. In your second run those pods would have `hostNetwork: true` and their IP wouldn't match the node's, which is a possible scenario depending the network config (see [comment above](https://github.com/linkerd/linkerd2/pull/3513#issuecomment-538145228)).

> Now that we explicitly handle node internal IP, will it still be possible for multiple running pods to have the same IP?

Multiple running pods with the same IP is indeed the case when they all have `hostNetwork: true`. 

> Is it bad to just issue a warning with the names of the pods if len(pods) > 0, and just return nil, nil?

What would the dashboard show in that case? Currently in this branch, if that happens and the IP matches the node, we just say "something in that node" is sending unmeshed traffic. If the IP doesn't match the node, we say "IP/1.2.3.4"

Reply:
> those pods would have hostNetwork: true and their IP wouldn't match the node's

Ah.. somehow k8s API sees it as the CNI gateway IP (e.g. in minikube), instead of the pod IP?

> Multiple running pods with the same IP is indeed the case when they all have hostNetwork: true.

If there are multiple pods sharing the same host IP, wouldn't the change in [`controller/tap/server.go`](https://github.com/linkerd/linkerd2/blob/9a4d97675303bfbf2db60fa75859cb847f2264a8/controller/tap/server.go#L619-L622) be `true` and return?

> What would the dashboard show in that case? 

Not sure tbh :stuck_out_tongue:. Ideally, it will be nice to separate pods using host network from say kubelet traffic. But yeah, we don't need to tackle this here. We can check with @grampelberg if we want to open a separate issue for it.

Reply:
I think we're getting about as good as we can with this method. It'll be interesting to flesh out something more comprehensive, but I think we need to look at it more deeply first.